operator/controllers/actionsrunner/reconciler.go

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright 2020 In Loco Tecnologia da Informação S.A.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actionsrunner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	inlocov1alpha1 "github.com/inloco/kube-actions/operator/api/v1alpha1"
+	"github.com/inloco/kube-actions/operator/controllers/actionsrunner/facades"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	patchOpts = []client.PatchOption{
+		client.ForceOwnership,
+		client.FieldOwner("kube-actions"),
+	}
+
+	deleteOpts = []client.DeleteOption{
+		client.PropagationPolicy(metav1.DeletePropagationForeground),
+	}
+)
+
+// Reconciler reconciles a ActionsRunner object
+type Reconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	source  chan event.GenericEvent
+	waiting sync.Map
+	wires   map[types.NamespacedName]*Wire
```

Comment:
Thought about this and figured that we may benefit from encapsulating sync objs and channels in separate structs...

Code written in this way may be easier to reason about since channels would not live outside these structs (though we indirectly use them to communicate via their struct's interface)

A suggestion:

```go
type Reconciler struct {
   ...
   wc *WireCollection
}

type WireCollection interface {
  FetchOrCreate(actionRunner *ar, adoEventsSource chan) *Wire
  Close(actionRunner *ar)
}

type Wire interface  {
  ConsumeMsg(
    requestFunc func() error,
    cancellationFunc func() error,
  )
}
```

Using closures as `ConsumeMsg` params may seem overkill at first sight but I see it as a common pattern to avoid passing out ack channels to inform completion.

Reply:
It is a good idea but I would create another struct to handle the sync.Map, as it is used to track goroutines that are waiting for the job completion and have nothing to do with the wire.